src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DataReaderTest/DataReaderTest.cs

```diff
@@ -288,5 +292,77 @@ public static void CheckHiddenColumns()
         }
+
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringsSetup))]
+        public static void CheckNullRowVersionIsBDNull()
+        {
+            lock (s_rowVersionLock)
+            {
+                bool? originalValue = SetLegacyRowVersionNullBehaviour(false);
+                try
+                {
+                    using (SqlConnection con = new SqlConnection(DataTestUtility.TCPConnectionString))
+                    {
+                        con.Open();
+                        using (SqlCommand command = con.CreateCommand())
+                        {
+                            command.CommandText = "select cast(null as rowversion) rv";
+                            using (SqlDataReader reader = command.ExecuteReader())
+                            {
+                                reader.Read();
+                                Assert.True(reader.IsDBNull(0));
+                                Assert.Equal(reader[0], DBNull.Value);
+                            }
+                        }
+                    }
+                }
+                finally
+                {
+                    SetLegacyRowVersionNullBehaviour(originalValue);
+                }
+            }
+        }
+
+        [ConditionalFact(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringsSetup))]
+        public static void CheckLegacyNullRowVersionIsEmptyArray()
+        {
+            lock (s_rowVersionLock)
+            {
+                bool? originalValue = SetLegacyRowVersionNullBehaviour(true);
+                try
+                {
+                    using (SqlConnection con = new SqlConnection(DataTestUtility.TCPConnectionString))
+                    {
+                        con.Open();
+                        using (SqlCommand command = con.CreateCommand())
+                        {
+                            command.CommandText = "select cast(null as rowversion) rv";
+                            using (SqlDataReader reader = command.ExecuteReader())
+                            {
+                                reader.Read();
+                                Assert.False(reader.IsDBNull(0));
+                                SqlBinary value = reader.GetSqlBinary(0);
+                                Assert.False(value.IsNull);
+                                Assert.Equal(0, value.Length);
+                                Assert.NotNull(value.Value);
+                            }
+                        }
+                    }
+                }
+                finally
+                {
+                    SetLegacyRowVersionNullBehaviour(originalValue);
+                }
+            }
+        }
+
+        private static bool? SetLegacyRowVersionNullBehaviour(bool? value)
+        {
+            Type switchesType = typeof(SqlCommand).Assembly.GetType("Microsoft.Data.SqlClient.LocalAppContextSwitches");
```

Comment:
Could you not simply call `AppContext.SetSwitch("Switch.Microsoft.Data.SqlClient.LegacyRowVersionNullBehaviour",true);` and change driver behavior ?
Any special reason to use reflection ?

Reply:
It's cached per process. Whichever test ran first would cause the other to fail. I have to set the value I want explicitly each time.

Reply:
You need to set/reset before and after running test to ensure guaranteed setting.

Reply:
No I don't because I'm setting the cache variable and totally bypassing the appcontext switch. Check the code carefully.